collections/iter.go

```diff
@@ -0,0 +1,312 @@
+	r.prefix = &key
+	return r
+}
+
+// StartInclusive makes the range contain only keys which are bigger or equal to the provided start K.
+func (r *Range[K]) StartInclusive(start K) *Range[K] {
+	r.start = BoundInclusive(start)
+	return r
+}
+
+// StartExclusive makes the range contain only keys which are bigger to the provided start K.
+func (r *Range[K]) StartExclusive(start K) *Range[K] {
+	r.start = BoundExclusive(start)
+	return r
+}
+
+// EndInclusive makes the range contain only keys which are smaller or equal to the provided end K.
+func (r *Range[K]) EndInclusive(end K) *Range[K] {
+	r.end = BoundInclusive(end)
+	return r
+}
+
+// EndExclusive makes the range contain only keys which are smaller to the provided end K.
+func (r *Range[K]) EndExclusive(end K) *Range[K] {
+	r.end = BoundExclusive(end)
+	return r
+}
+
+func (r *Range[K]) Descending() *Range[K] {
+	r.order = OrderDescending
+	return r
+}
+
+// test sentinel error
+var errRange = errors.New("collections: range error")
+
+func (r *Range[K]) RangeValues() (prefix *K, start *Bound[K], end *Bound[K], order Order, err error) {
+	if r.prefix != nil && (r.end != nil || r.start != nil) {
+		return nil, nil, nil, order, fmt.Errorf("%w: prefix must not be set if either start or end are specified", errRange)
+	}
+	return r.prefix, r.start, r.end, r.order, nil
+}
+
+// iteratorFromRanger generates an Iterator instance, with the proper prefixing and ranging.
+// a nil Ranger can be seen as an ascending iteration over all the possible keys.
+func iteratorFromRanger[K, V any](ctx context.Context, m Map[K, V], r Ranger[K]) (iter Iterator[K, V], err error) {
+	// defaults
+	var (
+		prefix *K
+		start  *Bound[K]
+		end    *Bound[K]
+		order  = OrderAscending
+	)
+	// override defaults only if a Ranger is provided.
+	if r != nil {
+		prefix, start, end, order, err = r.RangeValues()
+		if err != nil {
+			return iter, err
+		}
+	}
+	var prefixBytes []byte
+	if prefix != nil {
+		prefixBytes, err = encodeKeyWithPrefix(m.prefix, m.kc, *prefix)
+		if err != nil {
+			return iter, err
+		}
+	} else {
+		prefixBytes = m.prefix
+	}
+	var startBytes []byte // default is nil
+	if start != nil {
+		startBytes, err = encodeKeyWithPrefix(nil, m.kc, start.value)
+		if err != nil {
+			return
+		}
+		// iterators are inclusive at start by default
+		// so if we want to make the iteration exclusive
+		// we extend by one byte.
+		if !start.inclusive {
+			startBytes = extendOneByte(startBytes)
+		}
+	}
+	var endBytes []byte // default is nil
+	if end != nil {
+		endBytes, err = encodeKeyWithPrefix(nil, m.kc, end.value)
+		if err != nil {
+			return iter, err
+		}
+		// iterators are exclusive at end by default
+		// so if we want to make the iteration
+		// inclusive we need to extend by one byte.
+		if end.inclusive {
+			endBytes = extendOneByte(endBytes)
+		}
+	}
+
+	store, err := m.getStore(ctx)
+	if err != nil {
+		return iter, err
+	}
+	prefixedStartBytes := append(prefixBytes, startBytes...)
```

Comment:
I'd rather not do this type of appending. I think it should be either prefix or start, otherwise this function could create a corrupted prefix from a bad ranger implementation